File: iOS/Shiba/Shiba/ShibaFunctionExecutor.swift
//
// Created by 高垣朝陽 on 2018/8/16.
// Copyright (c) 2018 Tlaster. All rights reserved.
//

import Foundation

class ShibaFunctionExecutor {

    func execute(function: ShibaFunction, dataContext: Any?) -> Any? {
        return Shiba.instance.configuration.converterExecutor.execute(name: function.name, parameters: function.parameter.map { it in
            getParameterValue(parameter: it, dataContext: dataContext)
        })
    }

    private func getParameterValue(parameter: Any, dataContext: Any?) -> Any? {
        switch parameter {
        case is ShibaFunction:
            return execute(function: parameter as! ShibaFunction, dataContext: dataContext)
        case is ShibaBinding:
            return getValueFromDataContext(binding: parameter as! ShibaBinding, dataContext: dataContext)
        default:
            return parameter
        }
    }

    func getValueFromDataContext(binding: ShibaBinding, dataContext: Any?) -> Any? {
        return Shiba.instance.configuration.bindingValueResolver.getValue(dataContext: dataContext, name: binding.path)
    }
}

class SingleBindingShibaFunctionExecutor: ShibaFunctionExecutor {
    override func getValueFromDataContext(binding: ShibaBinding, dataContext: Any?) -> Any? {
        return dataContext;
    }
}

File: Android/shiba/src/main/java/moe/tlaster/shiba/ShibaFunctionExecutor.kt
package moe.tlaster.shiba

import moe.tlaster.shiba.type.ShibaExtension
import moe.tlaster.shiba.type.ShibaFunction

open class ShibaFunctionExecutor {
    fun execute(function: ShibaFunction, dataContext: Any?) : Any? {
        return if (Shiba.configuration.nativeConverter.containsKey(function.name)) {
            Shiba.configuration.nativeConverter[function.name]?.invoke(function.parameter.map { getParameterValue(it, dataContext) })
        } else {
            Shiba.configuration.scriptRuntime.callFunction(function.name, *function.parameter.map { getParameterValue(it, dataContext) }.toTypedArray())
        }
    }

    private fun getParameterValue(parameter: Any, dataContext: Any?) : Any? {
        return when (parameter) {
            is ShibaFunction -> execute(parameter, dataContext)
            is ShibaExtension -> getValueFromDataContext(parameter, dataContext)
            else -> parameter
        }
    }

    protected open fun getValueFromDataContext(extension: ShibaExtension, dataContext: Any?) : Any? {
        return Shiba.configuration.extensionExecutors.firstOrNull { it.name == extension.type }?.provideValue(null, extension)
    }
}

class SingleBindingShibaFunctionExecutor : ShibaFunctionExecutor() {
    override fun getValueFromDataContext(extension: ShibaExtension, dataContext: Any?): Any? {
        return dataContext
    }
}

